File: 卡码/代码随想录/JS/单调栈/LeetCode_0084.js
//84. 柱状图中最大的矩形
const largestRectangleArea = function(heights) {
	heights.unshift(0);
	heights.push(0);
	const st=new Array();
	let res=0;
	st.push(0);
	for(let i=1;i<heights.length;i++){
		if(heights[i]==heights[st[st.length-1]]) st.pop();
		else{
			while(heights[i]<heights[st[st.length-1]]){
				const mid=st.pop();
				if(st.length){
					const h=heights[mid];
					const w=i-st[st.length-1]-1;
					res=Math.max(res,h*w);
				}
			}
		}
		st.push(i);
	}
	return res;
};

File: 卡码/代码随想录/Go/单调栈/LeetCode_0084.go
// 84. 柱状图中最大的矩形
package main

import (
	"math"
)

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

func largestRectangleArea(heights []int) int {
	res := 0
	newHeights := make([]int, len(heights)+2)
	for i := 0; i < len(heights); i++ {
		newHeights[i+1] = heights[i]
	}
	st := make([]int, 0)
	for i, height := range newHeights {
		for len(st) > 0 && height < newHeights[st[len(st)-1]] {
			mid := st[len(st)-1]
			st = st[:len(st)-1]
			h := newHeights[mid]
			w := i - st[len(st)-1] - 1
			res = max(res, h*w)
		}
		st = append(st, i)
	}
	return res
}
